File: C11/백준/Silver/21736. 헌내기는 친구가 필요해/헌내기는 친구가 필요해.c
#include <stdio.h>

#define MAX 600

typedef struct {
    int x, y;
} Coord;

int N, M;
char map[MAX][MAX + 1];
int visited[MAX][MAX];

int dx[4] = {0, 0, -1, 1};
int dy[4] = {1, -1, 0, 0};

Coord queue[MAX * MAX];
int front = 0, rear = 0;

void enqueue(int x, int y) {
    queue[rear].x = x;
    queue[rear].y = y;
    rear++;
}

Coord dequeue() {
    return queue[front++];
}

int isEmpty() {
    return front == rear;
}

int bfs(int x, int y){
    enqueue(x, y);
    visited[x][y] = 1;
    int count = 0;

    while (!isEmpty()) {
        Coord current = dequeue();

        if (map[current.x][current.y] == 'P') {
            count++;
        }

        for (int i = 0; i < 4; i++) {
            int nx = current.x + dx[i];
            int ny = current.y + dy[i];

            if (nx >= 0 && ny >= 0 && nx < N && ny < M) {
                if (!visited[nx][ny] && map[nx][ny] != 'X') {
                    visited[nx][ny] = 1;
                    enqueue(nx, ny);
                }
            }
        }
    }

    return count;
}

int main(void){
    scanf("%d %d", &N, &M);
    getchar();

    int startX = -1, startY = -1;

    for (int i = 0; i < N; i++) {
        scanf("%s", map[i]);
        for (int j = 0; j < M; j++) {
            if (map[i][j] == 'I') {
                startX = i;
                startY = j;
            }
        }
    }

    int result = bfs(startX, startY);

    if (result == 0) {
        printf("TT");
    } else {
        printf("%d", result);
    }

    return 0;
}

File: Java/백준/Silver/21736. 헌내기는 친구가 필요해/헌내기는 친구가 필요해.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;

public class Main {

    static int N, M;
    static char[][] map;
    static boolean[][] visited;

    //상, 하, 좌, 우
    static int[] dx = {0, 0, -1, 1};
    static int[] dy = {1, -1, 0, 0};

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String[] campus = br.readLine().split(" ");
        N = Integer.parseInt(campus[0]);
        M = Integer.parseInt(campus[1]);

        map = new char[N][M];
        visited = new boolean[N][M];

        int startX = 0, startY = 0;

        for (int i = 0; i < N; i++) {
            String line = br.readLine();
            for (int j = 0; j < M; j++) {
                map[i][j] = line.charAt(j);
                if (map[i][j] == 'I') {
                    startX = i;
                    startY = j;
                }
            }
        }

        int result = bfs(startX, startY);
        System.out.print(result == 0 ? "TT" : result);
    }

    static int bfs(int x, int y) {
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[]{x, y});
        visited[x][y] = true;
        int count = 0;

        while (!q.isEmpty()) {
            int[] now = q.poll();
            int cX = now[0], cY = now[1];

            if (map[cX][cY] == 'P') { //사람을 만난 경우
                count++;
            }

            for (int i = 0; i < 4; i++) {
                int nx = cX + dx[i], ny = cY + dy[i];
                if (nx >= 0 && ny >= 0 && nx < N && ny < M) { //캠퍼스 밖으로 나갈 수 없음
                    if (!visited[nx][ny] && map[nx][ny] != 'X') { //벽으로는 이동할 수 없음
                        visited[nx][ny] = true;
                        q.offer(new int[]{nx, ny});
                    }
                }
            }
        }

        return count;
    }
}